Invalid error message for a line missing semicolon

I tried this code:

```rust
           find(line.to_string(), &cap[0]).iter().for_each(|x| {
                #[allow(clippy::cast_possible_truncation)]
                coordinates.push((*x as u16, i as u16));  
            })

            *line = replace;
```

I expected to see this error happen: Missing semicolon at the end of statement
Instead, these two errors happened: 
```
error[E0070]: invalid left-hand side of assignment
error[E0369]: cannot multiply `&mut String` to `()`
```

### Meta
I ran this on both stable and nightly. Here are the versions
```
rustc 1.48.0 (7eac88abb 2020-11-16)
binary: rustc
commit-hash: 7eac88abb2e57e752f3302f02be5f3ce3d7adfb4
commit-date: 2020-11-16
host: x86_64-unknown-linux-gnu
release: 1.48.0
LLVM version: 11.0

rustc 1.50.0-nightly (fe982319a 2020-11-19)
binary: rustc
commit-hash: fe982319aa0aa5bbfc2795791a753832292bd2ba
commit-date: 2020-11-19
host: x86_64-unknown-linux-gnu
release: 1.50.0-nightly
```

<details><summary>Complete error</summary>
<p>

```
error[E0070]: invalid left-hand side of assignment
   --> src/search.rs:150:19
    |
145 | /             find(line.to_string(), &cap[0]).iter().for_each(|x| {
146 | |                 #[allow(clippy::cast_possible_truncation)]
147 | |                 coordinates.push((*x as u16, i as u16));
148 | |             })
149 | |
150 | |             *line = replace;
    | |                 - ^
    | |_________________|
    |                   cannot assign to this expression

error[E0369]: cannot multiply `&mut String` to `()`
   --> src/search.rs:150:13
    |
145 | /             find(line.to_string(), &cap[0]).iter().for_each(|x| {
146 | |                 #[allow(clippy::cast_possible_truncation)]
147 | |                 coordinates.push((*x as u16, i as u16));
148 | |             })
    | |______________- ()
149 |
150 |               *line = replace;
    |               ^---- &mut String
```

</p>
</details>

I'm pretty sure rustc's lexer/parser discard all whitespace, including newlines, which unfortunately makes emitting a better diagnostic difficult, if not impossible.

Essentially, the parser sees this:
```rust
find(line.to_string(), &cap[0]).iter().for_each(...) * line = replace;
```
and reads it as an assignment to the result of a multiplication, which is why it issues these diagnostics. Unfortunately, this case is difficult to discern from cases where the diagnostics that rustc already emits would be correct:
```rust
fn main() {
    let x = 4;
    let y;
    // I wonder if Rust has a simple algebra solver...
    x * y = 8;
    assert_eq!(y, 2);
}
```

```
error[E0070]: invalid left-hand side of assignment
 --> src/main.rs:4:11
  |
4 |     x * y = 8;
  |     ----- ^
  |     |
  |     cannot assign to this expression
```

Whitespace probably cannot be made significant for multiplications at this point since that would be a breaking change. There are probably codebases that do something like this:
```rust
let foobar = very.long(expressions, that).take(up, a, whole).bunch::<of, SPACE>()
    * very.long(expressions, that).take(up, a, whole).bunch::<of, SPACE>()
    * very.long(expressions, that).take(up, a, whole).bunch::<of, SPACE>()
    * very.long(expressions, that).take(up, a, whole).bunch::<of, SPACE>();
```

Since the left hand side of an assignment can be arbitrarily long, discerning a multiplication broken across multiple lines from a statement with a missing semicolon followed by an assignment to a dereference would probably require the parser to look infinitely far ahead, which as far as I know is something that Rust actively avoids.

--- a/compiler/rustc_hir_typeck/src/expr.rs
+++ b/compiler/rustc_hir_typeck/src/expr.rs
@@ -953,10 +953,31 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 Applicability::MachineApplicable,
             );
         });
+        self.check_for_missing_semi(lhs, &mut err);
 
         adjust_err(&mut err);
 
         err.emit();
+    }
+
+    /// Check if the expression that could not be assigned to was a typoed expression that
+    fn check_for_missing_semi(&self, expr: &'tcx hir::Expr<'tcx>, err: &mut Diagnostic) {
+        if let hir::ExprKind::Binary(binop, lhs, rhs) = expr.kind
+            && let hir::BinOpKind::Mul = binop.node
+            && self.tcx.sess.source_map().is_multiline(lhs.span.between(rhs.span))
+            && rhs.is_syntactic_place_expr()
+        {
+            //      v missing semicolon here
+            // foo()
+            // *bar = baz;
+            // (#80446).
+            err.span_suggestion_verbose(
+                lhs.span.shrink_to_hi(),
+                "you might have meant to write a semicolon here",
+                ";".to_string(),
+                Applicability::MachineApplicable,
+            );
+        }
     }
 
     // Check if an expression `original_expr_id` comes from the condition of a while loop,

--- a/compiler/rustc_hir_typeck/src/expr.rs
+++ b/compiler/rustc_hir_typeck/src/expr.rs
@@ -961,7 +961,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     }
 
     /// Check if the expression that could not be assigned to was a typoed expression that
-    fn check_for_missing_semi(&self, expr: &'tcx hir::Expr<'tcx>, err: &mut Diagnostic) {
+    pub fn check_for_missing_semi(
+        &self,
+        expr: &'tcx hir::Expr<'tcx>,
+        err: &mut DiagnosticBuilder<'_, ErrorGuaranteed>,
+    ) -> bool {
         if let hir::ExprKind::Binary(binop, lhs, rhs) = expr.kind
             && let hir::BinOpKind::Mul = binop.node
             && self.tcx.sess.source_map().is_multiline(lhs.span.between(rhs.span))
@@ -977,7 +981,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 ";".to_string(),
                 Applicability::MachineApplicable,
             );
+            return true;
         }
+        false
     }
 
     // Check if an expression `original_expr_id` comes from the condition of a while loop,

--- a/compiler/rustc_hir_typeck/src/op.rs
+++ b/compiler/rustc_hir_typeck/src/op.rs
@@ -379,6 +379,22 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         (err, output_def_id)
                     }
                 };
+                if self.tcx.sess.source_map().is_multiline(lhs_expr.span.between(rhs_expr.span))
+                    && let IsAssign::No = is_assign
+                    && let hir::BinOpKind::Mul = op.node
+                    && rhs_expr.is_syntactic_place_expr()
+                {
+                    //      v missing semicolon here
+                    // foo()
+                    // *bar = baz;
+                    // (#80446).
+                    err.span_suggestion_verbose(
+                        lhs_expr.span.shrink_to_hi(),
+                        "you might have meant to write a semicolon here",
+                        ";".to_string(),
+                        Applicability::MaybeIncorrect,
+                    );
+                }
 
                 let suggest_deref_binop =
                     |err: &mut DiagnosticBuilder<'_, _>, lhs_deref_ty: Ty<'tcx>| {

--- a/compiler/rustc_hir_typeck/src/op.rs
+++ b/compiler/rustc_hir_typeck/src/op.rs
@@ -379,21 +379,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         (err, output_def_id)
                     }
                 };
-                if self.tcx.sess.source_map().is_multiline(lhs_expr.span.between(rhs_expr.span))
-                    && let IsAssign::No = is_assign
-                    && let hir::BinOpKind::Mul = op.node
-                    && rhs_expr.is_syntactic_place_expr()
+                if self.check_for_missing_semi(expr, &mut err)
+                    && let hir::Node::Expr(expr) = self.tcx.hir().get_parent(expr.hir_id)
+                    && let hir::ExprKind::Assign(..) = expr.kind
                 {
-                    //      v missing semicolon here
-                    // foo()
-                    // *bar = baz;
-                    // (#80446).
-                    err.span_suggestion_verbose(
-                        lhs_expr.span.shrink_to_hi(),
-                        "you might have meant to write a semicolon here",
-                        ";".to_string(),
-                        Applicability::MaybeIncorrect,
-                    );
+                    // We defer to the later error produced by `check_lhs_assignable`.
+                    err.delay_as_bug();
                 }
 
                 let suggest_deref_binop =

--- a/tests/ui/binop/false-binop-caused-by-missing-semi.fixed
+++ b/tests/ui/binop/false-binop-caused-by-missing-semi.fixed
@@ -3,7 +3,7 @@ fn foo() {}
 fn main() {
     let mut y = 42;
     let x = &mut y;
-    foo()
+    foo();
     *x = 0;  //~ ERROR invalid left-hand side of assignment
     let _ = x;
     println!("{y}");

--- a/tests/ui/binop/false-binop-caused-by-missing-semi.rs
+++ b/tests/ui/binop/false-binop-caused-by-missing-semi.rs
@@ -1,0 +1,9 @@
+fn foo() {}
+fn main() {
+    let mut y = 42;
+    let x = &mut y;
+    foo()
+    *x = 0;  //~ ERROR invalid left-hand side of assignment
+    //~^ ERROR cannot multiply `()` by `&mut {integer}`
+    println!("{y}");
+}

--- a/tests/ui/binop/false-binop-caused-by-missing-semi.stderr
+++ b/tests/ui/binop/false-binop-caused-by-missing-semi.stderr
@@ -1,18 +1,5 @@
-error[E0369]: cannot multiply `()` by `&mut {integer}`
-  --> $DIR/false-binop-caused-by-missing-semi.rs:6:5
-   |
-LL |     foo()
-   |     ----- ()
-LL |     *x = 0;
-   |     ^- &mut {integer}
-   |
-help: you might have meant to write a semicolon here
-   |
-LL |     foo();
-   |          +
-
 error[E0070]: invalid left-hand side of assignment
-  --> $DIR/false-binop-caused-by-missing-semi.rs:6:8
+  --> $DIR/false-binop-caused-by-missing-semi.rs:7:8
    |
 LL | /     foo()
 LL | |     *x = 0;
@@ -25,7 +12,6 @@ help: you might have meant to write a semicolon here
 LL |     foo();
    |          +
 
-error: aborting due to 2 previous errors
+error: aborting due to previous error
 
-Some errors have detailed explanations: E0070, E0369.
-For more information about an error, try `rustc --explain E0070`.
+For more information about this error, try `rustc --explain E0070`.

--- a/tests/ui/binop/false-binop-caused-by-missing-semi.stderr
+++ b/tests/ui/binop/false-binop-caused-by-missing-semi.stderr
@@ -1,0 +1,31 @@
+error[E0369]: cannot multiply `()` by `&mut {integer}`
+  --> $DIR/false-binop-caused-by-missing-semi.rs:6:5
+   |
+LL |     foo()
+   |     ----- ()
+LL |     *x = 0;
+   |     ^- &mut {integer}
+   |
+help: you might have meant to write a semicolon here
+   |
+LL |     foo();
+   |          +
+
+error[E0070]: invalid left-hand side of assignment
+  --> $DIR/false-binop-caused-by-missing-semi.rs:6:8
+   |
+LL | /     foo()
+LL | |     *x = 0;
+   | |      - ^
+   | |______|
+   |        cannot assign to this expression
+   |
+help: you might have meant to write a semicolon here
+   |
+LL |     foo();
+   |          +
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0070, E0369.
+For more information about an error, try `rustc --explain E0070`.

